src/Autofac/Diagnostics/DiagnosticTracerBase.cs

```diff
@@ -0,0 +1,367 @@
+        public virtual void OnOperationStart(OperationStartDiagnosticData data)
+        {
+        }
+
+        /// <summary>
+        /// Handles the event raised when a resolve operation completes successfully.
+        /// </summary>
+        /// <param name="data">
+        /// Diagnostic data associated with the event.
+        /// </param>
+        /// <remarks>
+        /// <para>
+        /// Derived classes can override this method and perform actions based
+        /// on the event. By default, the base class does nothing.
+        /// </para>
+        /// </remarks>
+        public virtual void OnOperationSuccess(OperationSuccessDiagnosticData data)
+        {
+        }
+
+        /// <summary>
+        /// Handles the event raised when a resolve request encounters an error.
+        /// </summary>
+        /// <param name="data">
+        /// Diagnostic data associated with the event.
+        /// </param>
+        /// <remarks>
+        /// <para>
+        /// Derived classes can override this method and perform actions based
+        /// on the event. By default, the base class does nothing.
+        /// </para>
+        /// </remarks>
+        public virtual void OnRequestFailure(RequestFailureDiagnosticData data)
+        {
+        }
+
+        /// <summary>
+        /// Handles the event raised when a resolve request starts.
+        /// </summary>
+        /// <param name="data">
+        /// Diagnostic data associated with the event.
+        /// </param>
+        /// <remarks>
+        /// <para>
+        /// Derived classes can override this method and perform actions based
+        /// on the event. By default, the base class does nothing.
+        /// </para>
+        /// </remarks>
+        public virtual void OnRequestStart(RequestDiagnosticData data)
+        {
+        }
+
+        /// <summary>
+        /// Handles the event raised when a resolve request completes successfully.
+        /// </summary>
+        /// <param name="data">
+        /// Diagnostic data associated with the event.
+        /// </param>
+        /// <remarks>
+        /// <para>
+        /// Derived classes can override this method and perform actions based
+        /// on the event. By default, the base class does nothing.
+        /// </para>
+        /// </remarks>
+        public virtual void OnRequestSuccess(RequestDiagnosticData data)
+        {
+        }
+
+        /// <summary>
+        /// Handles inbound events and converts the diagnostic data to a
+        /// strongly-typed object that can be handled by the other methods
+        /// in this observer.
+        /// </summary>
+        /// <param name="diagnosticName">
+        /// The name of the event that was raised. Diagnostic names are case-sensitive.
+        /// </param>
+        /// <param name="data">
+        /// The diagnostic data associated with the event.
+        /// </param>
+        public virtual void Write(string diagnosticName, object data)
```

Comment:
Torn on that because it isn't one of the Autofac events, it's sort of a pseudo-event. Adding "On" makes it seem like it's something that'd be raised by middleware or an operation. {I stole "Write" from the ASP.NET Core mechanism.)

Reply:
Understood, don't feel strongly about it. This becomes less of an issue if Write became protected.

Reply:
Left it `Write`, made it protected.